sdk/program/PROJECT.md

```diff
@@ -0,0 +1,138 @@
+## Future of the solana-program
+
+The Solana Program SDK exists in the Agave validator monorepo. This has been
+convenient for validator developers, but its presence in the monorepo has
+sometimes hindered on-chain program developers. Since its release cycle is tied
+to the Agave validator, it is not possible to easily release breaking changes.
+
+The issue for on-chain developers has been brought up or debated in issues like:
+
+* https://github.com/solana-labs/solana/issues/32503
+* https://github.com/solana-labs/solana/issues/32625
+* https://github.com/anza-xyz/agave/issues/20
+
+To give on-chain program developers a better experience, it's time to define
+some priorities and a plan to make things better.
+
+## Concepts for the Program SDK
+
+Here are the priorities for the future of the SDK, in order of importance.
+
+1. Secure
+2. Respect semantic versioning
+3. On-chain program developers come first
+4. Possible to build a new program quickly and easily
+5. Sensible defaults, but allow customization
+6. Minimal dependencies
+7. Kitchen sinks only re-export functionality
+8. Default to new crates or files to bundle functionality
+9. If new functionality can't be done in a new crate or file, use features
+10. Easy to publish
+11. Easy for Agave devs to make changes
+
+## Project Plan Phase 1: solana_program updates
+
+### Break it up: if it can reasonably be used independently, it gets its own crate
+
+As one approach to address point number 4 of easily allowing customization, we
+should avoid bundling too much functionality in solana-program, and instead
+break it up sensibly. This way, someone can easily drop in a custom entrypoint
+without pulling in the default during build time.
+
+This also helps address point number 8 of avoiding kitchen sinks. By putting
+only related functionality into separate crates, there is no temptation for
+developers to "just throw it into the SDK." Instead, they'll have to consider
+the best place for their changes.
+
+When breaking up solana-program, we will take it in phases. First, we'll extract
+functionality required for program clients, which typically involve creating
+instructions and deserializing types. Next, we will actually create program
+client crates for the native programs. And finally, we will extract functionality
+used for writing on-chain programs.
+
+Here's a more specific breakdown:
+
+* Add crates required for program clients:
+  * borsh
+  * clock
+  * example-mocks
+  * hash
+  * instruction
+  * message
```

Comment:
does `message` need to remain under the `program` sub-category of `sdk`?
It's not used by on-chain programs, right?

Reply:
I think `Message` was put there for on-chain signature verification, like secp256k1_recover, but it's a very specialized use-case

Reply:
but those just verify serialized bytes of a message (or just ixs iirc); doesn't need the message type itself, I think?

Reply:
Correct, it's not used directly, but if the signed bytes are a Solana message, a program might want to check the contents of the message. For example, if my program checks that the signed message bytes are an SPL token transfer, my program can create a message with one SPL token transfer and compare it against the signed message bytes.